Add Key Generation/import/export support for CryptoKeyOKP(x25519/ed25519) using CryptoKit

nmahendru opened the pull request:
#### 070a4485adadd784fdfc346dcbedf7d9b539cf6e
<pre>
Add Key Generation/import/export support for CryptoKeyOKP(x25519/ed25519) using CryptoKit
<a href="https://bugs.webkit.org/show_bug.cgi?id=280195">https://bugs.webkit.org/show_bug.cgi?id=280195</a>
<a href="https://rdar.apple.com/136368298">rdar://136368298</a>

Reviewed by David Kilzer.

Adding CryptoKit implementation of CryptoKeyOKP.
Serialization and Deserialization of SPKI/PKCS8 formats is not available
in CryptoKit so the the WebKit implementation is still used.

* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(EdKey.generatePrivateKey(_:)):
(EdKey.generatePrivateKeyKeyAgreement(_:)):
(EdKey.privateToPublic(_:priv:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: <a href="https://commits.webkit.org/284473@main">https://commits.webkit.org/284473@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7851eb3aac545004f3bf6dfb9f8c6561f185a069

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69507 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48907 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22180 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73591 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20665 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71624 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56708 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20516 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55266 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13727 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44605 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60013 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35745 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41271 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17443 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19042 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63210 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17789 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75302 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13489 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17003 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/62935 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13529 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60096 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62845 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15435 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10866 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4493 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44711 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45785 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46980 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45526 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

ddkilzer commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -476,6 +476,116 @@ public enum EdKeyAgreementAlgorithm {
+        }
+    }
+
+    public static func generatePrivateKeyKeyAgreement(algo: EdKeyAgreementAlgorithm) -> VectorUInt8
```
Nit:  Typo:  "...KeyKey..."?

If this a concatenation of "PrivateKey" and "KeyAgreement", you can ignore this comment.

nmahendru replied:
That is what it is `PrivateKey` + `KeyAgreement`

ddkilzer commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -476,6 +476,116 @@ public enum EdKeyAgreementAlgorithm {
+    public static func privateToPublic(algo: EdSigningAlgorithm, priv: SpanConstUInt8)
+        -> CryptoOperationReturnValue
+    {
+        var rv = CryptoOperationReturnValue()
```
Nit:  I'm not a fan of `rv` as a variable name.  Is it a return value?

If it's already used elsewhere in this file, it's okay to ignore this.

ddkilzer commented on Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
```diff
@@ -476,6 +476,116 @@ public enum EdKeyAgreementAlgorithm {
+        }
+    }
+
+    public static func privateToPublic(algo: EdSigningAlgorithm, priv: SpanConstUInt8)
```
Nit:  What is `priv` short for?  If `privateKey`, I'd suggest spelling it out.

Nit  Similarly, `algorithm` is better than `algo`.

If these are already used elsewhere in the file, ignore this feedback.

nmahendru replied:
Let me do a follow up with long names. I can just replace all of such instances with longer/specific names.

ddkilzer reviewed (approved):
r=me

nmahendru committed:
Add Key Generation/import/export support for CryptoKeyOKP(x25519/ed25519) using CryptoKit
https://bugs.webkit.org/show_bug.cgi?id=280195
rdar://136368298

Reviewed by David Kilzer.

Adding CryptoKit implementation of CryptoKeyOKP.
Serialization and Deserialization of SPKI/PKCS8 formats is not available
in CryptoKit so the the WebKit implementation is still used.

* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/WebCryptoAPI/import_export/okp_importKey.https.any.worker-expected.txt:
* Source/WebCore/PAL/pal/PALSwift/CryptoKitShim.swift:
(EdKey.generatePrivateKey(_:)):
(EdKey.generatePrivateKeyKeyAgreement(_:)):
(EdKey.privateToPublic(_:priv:)):
* Source/WebCore/PAL/pal/PALSwift/UnsafeOverlays.swift:
* Source/WebCore/crypto/cocoa/CryptoKeyOKPCocoa.cpp:
(WebCore::CryptoKeyOKP::platformGeneratePair):
(WebCore::CryptoKeyOKP::platformCheckPairedKeys):
(WebCore::CryptoKeyOKP::generateJwkX const):

Canonical link: https://commits.webkit.org/284473@main
